libcontainer/rootfs_linux.go

```diff
@@ -594,6 +595,37 @@ func createDevices(config *configs.Config) error {
 }
 
 func bindMountDeviceNode(dest string, node *configs.Device) error {
+	var st unix.Stat_t
+	if err := unix.Lstat(dest, &st); err != nil {
+		if !os.IsNotExist(err) {
+			return errors.Wrapf(err, "could not stat existing path for device node: %q", dest)
+		}
+	} else if os.FileMode(st.Mode)&os.ModeSymlink == os.ModeSymlink {
+		// If the path is a symlink, we remove it.
+		if err := unix.Unlink(dest); err != nil {
+			return errors.Wrapf(err, "device node %q already exists and is a symlink", dest)
+		}
+	} else {
+		switch os.FileMode(st.Mode) & os.ModeType {
+		case os.ModeDevice:
+			if node.Type != configs.BlockDevice {
+				return errors.Errorf("device node %q already exists but incorrect type %v (wanted block type)", dest, node.Type)
+			}
+		case os.ModeDevice | os.ModeCharDevice:
+			if node.Type != configs.CharDevice {
+				return errors.Errorf("device node %q already exists but incorrect type %v (wanted block type)", dest, node.Type)
+			}
+		default:
+			// NOP, happens on rootless
+		}
```

Comment:
This doesn't really make sense (and is why I changed it in my suggestion) -- what you're saying here is "if it's any other kind of inode, ignore it unless `node.Major` and `node.Minor` are non-zero because if the inode isn't a device then it won't have a major/minor number". I don't understand why you would want to do that.

The actual bug (and it was unrelated to rootless containers) was that I forgot that `os.Mode*` has different values from the Linux ones. This fix should've actually looked like:

```go
var st unix.Stat_t
if err := unix.Lstat(dest, &st); err != nil {
	if !os.IsNotExist(err) {
		return errors.Wrapf(err, "could not stat existing path for device node: %q", dest)
	}
} else if st.Mode&unix.S_IFMT == unix.S_IFLNK {
	// If the path is a symlink, we remove it.
	// NOTE: These semantics might be incorrect, but I'm not convinced
	//       that the semantics in the original PR are correct either.
	//       Resolving symlinks is a pretty dangerous game and I'd
	//       prefer not to do it if possible.
	if err := unix.Unlink(dest); err != nil {
		return errors.Wrapf(err, "device node %q already exists and is a symlink", dest)
	}
} else {
	// If there already exists a device inode with the correct (type,
	// major, minor), we can skip over setting it up. This logic only
	// really makes sense for rootless containers.
	var gotType configs.DeviceType
	switch st.Mode & unix.S_IFMT {
	case unix.S_IFBLK:
		gotType = configs.BlockDevice
	case unix.S_IFCHR:
		gotType = configs.CharDevice
	default:
		return errors.Errorf("device node %q already exists but not a valid device: %#o", dest, st.Mode)
	}
	gotMajor := int64(unix.Major(st.Rdev))
	gotMinor := int64(unix.Minor(st.Rdev))
	if gotType != node.Type || gotMajor != node.Major || gotMinor != node.Minor {
		return errors.Errorf("device node %q already exists but incorrect type [%s %d:%d] (wanted [%s %d:%d])", dest, gotType, gotMajor, gotMinor, node.Type, node.Major, node.Minor)
	}
	return nil
}
```

And you should change back the rest of the hunk too. (This version of the change is also incorrect because of the above misunderstanding -- it will effectively always ignore the type of the device and only compare the major:minor numbers because the bits in the switch cases are incorrect.)

Reply:
updated